### ISSUE TITLE
Retire stats.mtime in favor of MD5 hashes for static asset diffing

### ISSUE THREAD

**Issue by ryanblock** [architect/architect#890]
**Describe the issue**
Dating back [quite some ways](https://github.com/architect/architect/commit/5f45293c264967edcb27ff4437f8040a2b0b4726), we have relied on file stats `mtime` compared to S3's last updated time to know whether a file was modified. This has been surprisingly solid, despite the possibility of being brittle.

However, recently we began observing strange mtime mutation in Lambda when working with files, causing false negatives for modification. (Specifically, we're seeing files' `mtime` be unpredictably mutated, likely by the common [`yauzl` library](https://github.com/thejoshwolfe/yauzl).)

A reason not to have used S3's `ETag` headObject parameter (an MD5 hash) is because, as I recall, files used to be hashed in 5MB or 8MB chunks. (I believe this knowledge wasn't state secret, but it wasn't published, either. But it's been a while, and I may be wrong!) In any event, while this may still be the case, I found in testing today that >100MB binary files uploaded via SDK had fully intact MD5 hashes.

Additionally, I found that files uploaded via the web interface may be subject to non-standard hashing. But since we're talking about SDK-based uploads of files intended to be consumed by an application's end users, that should cover almost all Architect use cases I can think of.

As such, I'm going to begin the process of switching deploy over to using MD5 hashes instead of file stats, which may be nondeterministic.

**Comment by ryanblock** [comment]
Patched in https://github.com/architect/deploy/pull/257/files

**Comment by ryanblock** [comment]
Released in `6.5.2`. Quick update to memorialize some additional information I found re-researching this.

A lot of what I remembered was generally correct, but it's a lot clearer now that I found some better documentation, which wasn't around the last time I looked into this a few years back. In my own testing / research, the web interface uploads of large assets are via multipart, and it had a chunked hash, which comports with the reference: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html

Summary:

> The entity tag is a hash of the object. The ETag reflects changes only to the contents of an object, not its metadata. The ETag may or may not be an MD5 digest of the object data. Whether or not it is depends on how the object was created and how it is encrypted as described below:
>
> Objects created by the PUT Object, POST Object, or Copy operation, or through the AWS Management Console, and are encrypted by SSE-S3 or plaintext, have ETags that are an MD5 digest of their object data.
> 
> Objects created by the PUT Object, POST Object, or Copy operation, or through the AWS Management Console, and are encrypted by SSE-C or SSE-KMS, have ETags that are not an MD5 digest of their object data.
> 
> If an object is created by either the Multipart Upload or Part Copy operation, the ETag is not an MD5 digest, regardless of the method of encryption.

tl;dr – as long as the files are being put via SDK calls (and as long as Architect is publishing the assets), our asset hashes should be deterministic.
